src/System.Private.CoreLib/shared/System/Marvin.cs

```diff
@@ -28,73 +28,194 @@ internal static partial class Marvin
+
+            p0 += partialResult;
             Block(ref p0, ref p1);
             Block(ref p0, ref p1);
 
             return (int)(p1 ^ p0);
+
+        InputTooSmallToEnterMainLoop:
+
+            // We can't run the main loop, but we might still have 4 or more bytes available to us.
+            // If so, jump to the 4 .. 7 bytes logic immediately after the main loop.
+
+            if ((count & 4) != 0)
+            {
+                goto Between4And7BytesRemain;
+            }
+
+            // We had only 0 .. 3 bytes to begin with, so we can't perform any 32-bit reads.
+            // This means that we're going to be building up the final result right away and
+            // will only ever run two rounds total of the block function. Let's initialize
+            // the partial result to "no data".
+
+            if (BitConverter.IsLittleEndian)
+            {
+                partialResult = 0x80u;
+            }
+            else
+            {
+                partialResult = 0x80000000u;
+            }
+
+            if ((count & 1) != 0)
+            {
+                // If the buffer is 1 or 3 bytes in length, let's read a single byte now
+                // and merge it into our partial result. This will result in partialResult
+                // having one of the two values below, where AA BB CC are the buffer bytes.
+                // 
+                //                  (little-endian / big-endian)
+                // [ AA          ]  -> 0x000080AA / 0xAA800000;
+                // [ AA BB CC    ]  -> 0x000080CC / 0xCC800000;
+
+                partialResult = Unsafe.AddByteOffset(ref data, (nuint)(uint)count & 2);
+
+                if (BitConverter.IsLittleEndian)
+                {
+                    partialResult |= 0x8000;
+                }
+                else
+                {
+                    partialResult <<= 24;
+                    partialResult |= 0x800000u;
+                }
+            }
+
+            if ((count & 2) != 0)
+            {
+                // If the buffer is 2 or 3 bytes in length, let's read a single ushort now
+                // and merge it into the partial result. This will result in partialResult
+                // having one of the two values below, where AA BB CC are the buffer bytes.
+                // 
+                //                  (little-endian / big-endian)
+                // [ AA BB       ]  -> 0x0080BBAA / 0xAABB8000;
+                // [ AA BB CC    ]  -> 0x80CCBBAA / 0xAABBCC80; (carried over from above)
+
+                if (BitConverter.IsLittleEndian)
+                {
+                    partialResult <<= 16;
+                    partialResult |= (uint)Unsafe.ReadUnaligned<ushort>(ref data);
+                }
+                else
+                {
+                    partialResult |= (uint)Unsafe.ReadUnaligned<ushort>(ref data);
+                    partialResult = BitOps.RotateLeft(partialResult, 16);
+                }
+            }
```

Comment:
Does this end up generating equivalent tier 1 code to if the endianness checks were moved outside, e.g.
```C#
if (BitConverter.IsLittleEndian)
{
    partialResult = 0x80u;
    if ((count & 1) != 0) partialResult = Unsafe.AddByteOffset(ref data, (nuint)(uint)count & 2) | 0x8000;
    if ((count & 2) != 0) partialResult = (partialResult << 16) | (uint)Unsafe.ReadUnaligned<ushort>(ref data);
}
else
{
    ...
}
```
?

Reply:
It should. We prune those branches from the IL before shipping System.Private.CoreLib.

Reply:
BTW, are you suggesting you like a single outer `IsLittleEndian` or a bunch of inner `IsLittleEndian` checks? I inverted this to see what it would look like (endianness checks on the outside, bit mask checks on the inside), and the result makes the code more difficult to read IMO.

Reply:
> We prune those branches from the IL before shipping System.Private.CoreLib.

That is not correct. These branches will be in IL in the shipping CoreLib.

Reply:
Huh, that's unexpected. I thought we had a separate pass that did tree pruning of the IL for things like `IntPtr.Size` and `BitConverter.IsLittleEndian` since we know those values at compile time.

Regardless, I checked the codegen and it appears to be optimal. I don't see any unexpected branching.